src/firefly/java/edu/caltech/ipac/firefly/server/visualize/hips/HiPSMasterList.java

```diff
@@ -81,14 +87,25 @@ public class HiPSMasterList extends EmbeddedDbProcessor {
     static {
         sources.put(ServerParams.IRSA, new IrsaHiPSListSource());
         sources.put(ServerParams.CDS, new CDSHiPSListSource());
-        sources.put(ServerParams.EXTERNAL, new ExternalHiPSListSource());
         sources.put(ServerParams.LSST, new LsstHiPSListSource());
     }
 
     private static final Logger.LoggerImpl _log= Logger.getLogger();
     private static final String errMsg = "HiPS Map search: no HiPS maps found";
 
     public DataGroup fetchDataGroup(TableServerRequest request) throws DataAccessException {
+
+        String sourceUrl= request.getParam(HIPS_LIST_SOURCE);
+        String sourceUrlName= request.containsParam(HIPS_LIST_SOURCE_NAME) ?
+                request.getParam(HIPS_LIST_SOURCE_NAME) : sourceUrl;
+        if (sourceUrl != null && !sources.containsKey(sourceUrl)) {
+            sources.put(sourceUrlName, new URLHiPSListSource(sourceUrl,sourceUrl) );
+        }
+        if (request.getBooleanParam(ENSURE_SOURCE)) {
+            return createDataGroupFromSources();
+        }
```

Comment:
The API for this processor seems unclear or tangled to me.  It looks like you have a master(public) list of `sources` that can grow on demand but cannot shrink.  Additionally, it handles both `PUT` and `GET` in one operation which made it more confusing.
`ADD_HOC_SOURCE` is poorly named.  Maybe it meant to be `ADHOC_SOURCES` or `ADD_ADHOC_SOURCES`, but it is actually some sort of IVOA IDs that when given may affect the source as well as the filtering part of `GET`.  From the name, if it's `ADD`, then I would expect to see it added to `sources`(master), but I don't see it.

With `ENSURE_SOURCE` added, you may `PUT`(add) one additional source into the list, but then return the latest list as is, ignore everything else(filtering, sorting, prioritization).

Keep in mind, Firefly has a multi-user backend.  Imagine multiple notebook users loading(`PUT`) whatever they want, but now it is shared with or leaked into other sessions.

I don't have a suggestion; I'm just pointing out what I see.

Reply:
I can change `ADD_HOC_SOURCE` but you understand it has been like that for years. I will make a betters name (after I figure out what the parameters does), it is an easy change. 

ENSURE_SOURCE is meant to be a different mode. I could have made a new processor but that seems like overkill. It is not a big deal to have many new sources names added since it does not search them until requested. I think they will support multi usage just fine. The request was for to be able to dynamically hit hips list servers. This approach is very low overhead. In the real world, multiple clients will call ENSURE_SOURCE with the same parameters so it will do nothing except return a table of sources.

Also a hips list servers is not meant to be secret and there will never be many. You would not be specifying this from a notebook. It would be via the firefly url api.  If the approach needs to me more advanced we can update it in another round. However, right not it seems like a simpler approach is a good way to start.

I does not have to return anything. I will take that out. I mostly had it return the table of sources for debugging.